File: src/knowledge-ingestion/parsers.ts
import { createRequire } from 'module';
const requireCjs = createRequire(import.meta.url);
const pdf = requireCjs('pdf-parse');

import mammoth from 'mammoth';
import * as XLSX from 'xlsx';
import { DocumentParser } from '../shared/interfaces';
import { CanonicalDocument, DocumentSection } from '../shared/types';

export class PlainTextAndMarkdownParser implements DocumentParser {
  supports(fileType: string): boolean {
    const ext = fileType.toLowerCase();
    return ext === 'txt' || ext === 'md' || ext === 'markdown' || ext === 'text';
  }

  async parse(filePath: string, buffer: Buffer, fileName: string): Promise<CanonicalDocument> {
    const text = buffer.toString('utf8');
    const sections: DocumentSection[] = [];
    
    // Simple markdown parsing to extract headings vs paragraph structure
    const lines = text.split('\n');
    let currentSection: DocumentSection | null = null;
    let paragraphIndex = 1;

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;

      // Check for markdown headings (# Heading, ## Heading, etc.)
      const headingMatch = trimmed.match(/^(#{1,6})\s+(.+)$/);
      if (headingMatch) {
        const headingLevel = headingMatch[1].length;
        const title = headingMatch[2];
        
        const newSection: DocumentSection = {
          id: `sec_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`,
          title,
          type: 'section',
          content: '',
          children: []
        };
        
        sections.push(newSection);
        currentSection = newSection;
      } else {
        const para: DocumentSection = {
          id: `para_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`,
          type: 'paragraph',
          content: trimmed
        };

        if (currentSection) {
          currentSection.children = currentSection.children || [];
          currentSection.children.push(para);
        } else {
          // Top-level paragraph before any headings
          sections.push(para);
        }
      }
    }

    return {
      documentId: '', // Filled in by ingestion worker
      workspaceId: '', // Filled in by ingestion worker
      name: fileName,
      fileType: 'markdown',
      sections,
      metadata: {}
    };
  }
}

export class PdfDocumentParser implements DocumentParser {
  supports(fileType: string): boolean {
    return fileType.toLowerCase() === 'pdf';
  }

  async parse(filePath: string, buffer: Buffer, fileName: string): Promise<CanonicalDocument> {
    const parser = new pdf.PDFParse({ data: new Uint8Array(buffer) });
    const parsedPdf = await parser.getText();
    const text = parsedPdf.text || '';
    await parser.destroy();
    
    // Normalize newlines from PDF physical layout to form clean semantic paragraphs
    const lines = text.split('\n').map(line => line.trim());
    let normalizedText = '';
    for (const line of lines) {
      if (line === '') {
        normalizedText += '\n\n';
      } else {
        normalizedText += (normalizedText.endsWith('\n\n') || normalizedText === '') ? line : ' ' + line;
      }
    }

    const paragraphs = normalizedText
      .split(/\n\s*\n/)
      .map(p => p.trim())
      .filter(Boolean);

    const sections: DocumentSection[] = paragraphs.map((paraText, idx) => ({
      id: `pdf_p_${idx}_${Date.now()}`,
      type: 'paragraph',
      content: paraText,
      source: { page: 1 } // pdf-parse extracts continuous text, page numbers can be refined if needed
    }));

    return {
      documentId: '',
      workspaceId: '',
      name: fileName,
      fileType: 'pdf',
      sections,
      metadata: {
        numPages: parsedPdf.numpages,
        info: parsedPdf.info
      }
    };
  }
}

export class DocxDocumentParser implements DocumentParser {
  supports(fileType: string): boolean {
    return fileType.toLowerCase() === 'docx';
  }

  async parse(filePath: string, buffer: Buffer, fileName: string): Promise<CanonicalDocument> {
    const result = await mammoth.extractRawText({ buffer });
    const text = result.value || '';
    
    const paragraphs = text
      .split(/\n\s*\n/)
      .map(p => p.trim())
      .filter(Boolean);

    const sections: DocumentSection[] = paragraphs.map((paraText, idx) => ({
      id: `docx_p_${idx}_${Date.now()}`,
      type: 'paragraph',
      content: paraText
    }));

    return {
      documentId: '',
      workspaceId: '',
      name: fileName,
      fileType: 'docx',
      sections,
      metadata: {}
    };
  }
}

export class ExcelAndCsvParser implements DocumentParser {
  supports(fileType: string): boolean {
    const ext = fileType.toLowerCase();
    return ext === 'xlsx' || ext === 'xls' || ext === 'csv';
  }

  async parse(filePath: string, buffer: Buffer, fileName: string): Promise<CanonicalDocument> {
    const workbook = XLSX.read(buffer, { type: 'buffer' });
    const sections: DocumentSection[] = [];

    for (const sheetName of workbook.SheetNames) {
      const sheet = workbook.Sheets[sheetName];
      // Convert sheet to JSON rows
      const rows: any[] = XLSX.utils.sheet_to_json(sheet, { defval: '' });

      if (rows.length === 0) continue;

      // Extract columns / headers
      const columns = Object.keys(rows[0]);

      const sheetSection: DocumentSection = {
        id: `sheet_${sheetName}_${Date.now()}`,
        title: `Sheet: ${sheetName}`,
        type: 'section',
        content: `Excel Sheet: ${sheetName}. Columns: ${columns.join(', ')}`,
        children: [],
        source: { sheet: sheetName }
      };

      // Each row forms a table record paragraph
      rows.forEach((row, rowIdx) => {
        const rowText = Object.entries(row)
          .map(([key, val]) => `${key}: ${val}`)
          .join('\n');
        
        sheetSection.children?.push({
          id: `row_${sheetName}_${rowIdx}_${Date.now()}`,
          type: 'table',
          content: rowText,
          source: { sheet: sheetName, row: rowIdx + 1 }
        });
      });

      sections.push(sheetSection);
    }

    return {
      documentId: '',
      workspaceId: '',
      name: fileName,
      fileType: fileName.endsWith('.csv') ? 'csv' : 'xlsx',
      sections,
      metadata: {
        sheets: workbook.SheetNames
      }
    };
  }
}

export class DocumentParserRouter {
  private parsers: DocumentParser[];

  constructor() {
    this.parsers = [
      new PlainTextAndMarkdownParser(),
      new PdfDocumentParser(),
      new DocxDocumentParser(),
      new ExcelAndCsvParser()
    ];
  }

  async parse(filePath: string, buffer: Buffer, fileName: string): Promise<CanonicalDocument> {
    const ext = fileName.split('.').pop() || '';
    const parser = this.parsers.find(p => p.supports(ext));
    if (!parser) {
      throw new Error(`Unsupported file type: ${ext} for file: ${fileName}`);
    }
    return parser.parse(filePath, buffer, fileName);
  }
}

File: backend/app/services/ingestion/parsers.py
from __future__ import annotations

import math
import os
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


@dataclass
class DocumentSection:
    id: str
    title: str | None = None
    type: str = "paragraph"
    content: str = ""
    children: list["DocumentSection"] = field(default_factory=list)
    source: dict[str, Any] = field(default_factory=dict)


@dataclass
class CanonicalDocument:
    document_id: str
    workspace_id: str
    name: str
    file_type: str
    sections: list[DocumentSection]
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class DocumentChunk:
    workspace_id: str
    document_id: str
    document_name: str
    section_path: list[str]
    chunk_index: int
    content: str
    metadata: dict[str, Any]


def _parse_text_markdown(buffer: bytes, file_name: str) -> CanonicalDocument:
    text = buffer.decode("utf-8", errors="replace")
    sections: list[DocumentSection] = []
    current_section: DocumentSection | None = None

    for line in text.splitlines():
        trimmed = line.strip()
        if not trimmed:
            continue
        heading_match = re.match(r"^(#{1,6})\s+(.+)$", trimmed)
        if heading_match:
            current_section = DocumentSection(
                id=f"sec_{len(sections)}",
                title=heading_match.group(2),
                type="section",
                content="",
            )
            sections.append(current_section)
        else:
            para = DocumentSection(id=f"para_{len(sections)}", type="paragraph", content=trimmed)
            if current_section is not None:
                current_section.children.append(para)
            else:
                sections.append(para)

    return CanonicalDocument("", "", file_name, "markdown", sections)


def _parse_pdf(buffer: bytes, file_name: str) -> CanonicalDocument:
    from pypdf import PdfReader
    from io import BytesIO

    reader = PdfReader(BytesIO(buffer))
    text_parts = []
    for page in reader.pages:
        text_parts.append(page.extract_text() or "")
    text = "\n".join(text_parts)
    paragraphs = [p.strip() for p in re.split(r"\n\s*\n", text) if p.strip()]
    sections = [
        DocumentSection(id=f"pdf_{idx}", type="paragraph", content=para, source={"page": 1})
        for idx, para in enumerate(paragraphs)
    ]
    return CanonicalDocument("", "", file_name, "pdf", sections, {"numPages": len(reader.pages)})


def _parse_docx(buffer: bytes, file_name: str) -> CanonicalDocument:
    from io import BytesIO
    from docx import Document

    doc = Document(BytesIO(buffer))
    paragraphs = [p.text.strip() for p in doc.paragraphs if p.text.strip()]
    sections = [
        DocumentSection(id=f"docx_{idx}", type="paragraph", content=para)
        for idx, para in enumerate(paragraphs)
    ]
    return CanonicalDocument("", "", file_name, "docx", sections)


def _parse_spreadsheet(buffer: bytes, file_name: str) -> CanonicalDocument:
    from io import BytesIO
    from openpyxl import load_workbook

    workbook = load_workbook(BytesIO(buffer), read_only=True, data_only=True)
    sections: list[DocumentSection] = []
    for sheet_name in workbook.sheetnames:
        sheet = workbook[sheet_name]
        rows = list(sheet.iter_rows(values_only=True))
        if not rows:
            continue
        headers = [str(cell or "") for cell in rows[0]]
        sheet_section = DocumentSection(
            id=f"sheet_{sheet_name}",
            title=f"Sheet: {sheet_name}",
            type="section",
            content=f"Excel Sheet: {sheet_name}. Columns: {', '.join(headers)}",
            source={"sheet": sheet_name},
        )
        for row_idx, row in enumerate(rows[1:], start=1):
            row_text = "\n".join(f"{headers[i]}: {row[i]}" for i in range(min(len(headers), len(row))))
            sheet_section.children.append(
                DocumentSection(
                    id=f"row_{sheet_name}_{row_idx}",
                    type="table",
                    content=row_text,
                    source={"sheet": sheet_name, "row": row_idx},
                )
            )
        sections.append(sheet_section)
    return CanonicalDocument("", "", file_name, Path(file_name).suffix.lstrip("."), sections)


def parse_document(storage_path: str, file_name: str) -> CanonicalDocument:
    buffer = Path(storage_path).read_bytes()
    ext = Path(file_name).suffix.lower().lstrip(".")

    if ext in {"txt", "md", "markdown", "text"}:
        return _parse_text_markdown(buffer, file_name)
    if ext == "pdf":
        return _parse_pdf(buffer, file_name)
    if ext == "docx":
        return _parse_docx(buffer, file_name)
    if ext in {"xlsx", "xls", "csv"}:
        if ext == "csv":
            text = buffer.decode("utf-8", errors="replace")
            lines = [line.strip() for line in text.splitlines() if line.strip()]
            sections = [
                DocumentSection(id=f"csv_{idx}", type="paragraph", content=line)
                for idx, line in enumerate(lines)
            ]
            return CanonicalDocument("", "", file_name, "csv", sections)
        return _parse_spreadsheet(buffer, file_name)

    raise ValueError(f"Unsupported file type: {ext} for file {file_name}")
